### ISSUE TITLE
Mixing conditional and non-conditional google_project_iam_binding resources for the same role

### ISSUE THREAD

**Issue by gokeefe** [hashicorp/terraform-provider-google#9426]
The documentation for [`google_project_iam`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam) resources states:

```
google_project_iam_binding resources can be used in conjunction with google_project_iam_member resources only if they do not grant privilege to the same role.
```

The [Method: projects.getIamPolicy](https://cloud.google.com/resource-manager/reference/rest/v1/projects/getIamPolicy) endpoint returns an object where the role for a conditional IAM has a unique name. Example:

```
{'members': ['user:jdoe@example.com', 'user:jsmith@example.com'],
 'role': 'roles/compute.admin'},
{'members': ['user:alincoln@example.com'],
 'role': 'roles/compute.admin_withcond_a0bfd75561b99825fcfb'},
```

In my testing, I was able to apply a non-conditional `google_project_iam_binding` resource to add jdoe@example.com with compute.admin, and a `google_project_iam_member` resource with IAM conditions to add alincoln@example.com with the same role. I then added jsmith@example.com to the non-conditional `google_project_iam_binding` resource and applied the change and the conditional resource was unaffected. I performed the same tested with two `google_project_iam_binding` resources and the same role, one conditional and one not, and it appeared to work just fine.

So, to me, it would appear you actually can mix `google_project_iam_binding` and `google_project_iam_member` resources targeting the same role, as long as only one of them is non-conditional?

**Comment by edwardmedia** [comment]
@gokeefe I am not sure what you meant about **conditional** or **non-conditional**. Can you explain?

Most IAM resources provide three formats like **xxxx_iam_policy**, **xxxx_iam_binding** and **xxxx_iam_member**.  Please pay attention to the differences among them. In general, **xxxx_iam_policy** will remove ALL the existing settings, replaced by ones you provided in the resource. On the other side, both **xxxx_iam_binding** and **xxxx_iam_member** will NOT delete existing ones. You can think of them as actions of **create** and **update**. The difference between these two is one provide a `batch` (note  it is a list type for `member**s**) and the other is for individual member (a single string type). Because these two work on create/update, definitely you should avoid letting them to work on the same item.  Does this make sense?

**Comment by gokeefe** [comment]
Hi @edwardmedia, I mean a project_iam_binding or project_iam_member resource with a condition block ([IAM Condition](https://cloud.google.com/iam/docs/conditions-overview)) vs a resource without a condition block.

I'm aware of the differences based on the documentation, it just appears to be that they don't conflict if two resources target the same role, but one has conditions and the other doesn't, and the results from the getIamPolicy API method would appear to reflect that they are managed differently. Just wanted to try to verify that this assumption and testing results are correct.

**Comment by gokeefe** [comment]
With two google_project_iam_binding resources targeting the same role and no IAM Conditions, the conflicts are very noticable. Every time you run a terraform plan, it wants to change one or the other. But if one has conditions, and the other doesn't, no conflicts seem to occur.

**Comment by edwardmedia** [comment]
This is a good use case. @rileykarson what do you think?

**Comment by rileykarson** [comment]
That note in the documentation is a bit of a lie- the actual delimiter that would cause issues is if a member and binding both attempted to manage the same {role, condition} pair. I think the note predates conditions support in Terraform, honestly.

**Comment by gokeefe** [comment]
That makes sense. I _think_ though that the same applies to using more than one `google_project_iam_binding` resource as well, as long as each google_project_iam_binding resource targets a different {role, condition} (or lack of condition) pair. At least, that's how it appears in my testing.

**Comment by rileykarson** [comment]
That's correct, yeah, two bindings need to manage the same role+condition pair to conflict. Labelling as documentation for us to update.

**Comment by gokeefe** [comment]
Awesome, thank you for confirming!

**Comment by rileykarson** [comment]
note: m instead of xs to cover handwritten resources

**Comment by froblesmartin** [comment]
The problem with `google_project_iam_binding` is that it uses only the role as the internal Terraform ID of the resource. That is why, if you create two resources to bind the same role, they will collide. Will this be fixed when working on this issue, or should we create a new one for that specific thing? Maybe it is decided to be like that on purpose? 🤔

**Comment by gokeefe** [comment]
I don't think that's an issue; the documentation is pretty clear that a
`google_project_iam_binding` resource is authoritative for a given role, so
trying to define more than one unconditional IAM binding for the same role,
or a mix of unconditional iam policy, iam binding, or iam member resources,
will conflict.
